csharp/LogicalColumnReader.cs

```diff
@@ -286,36 +286,47 @@ public override int ReadBatch(Span<TElement> destination)
 
         private static Array ReadArrayLeafLevel(Node node, BufferedReader<TLogical, TPhysical> valueReader, short repetitionLevel, short definitionLevel)
         {
-            var values = new List<TLogical>();
-            var firstValue = true;
+            var valueChunks = new List<TLogical[]>();
             var innerNodeIsOptional = node.Repetition == Repetition.Optional;
             definitionLevel += (short) (innerNodeIsOptional ? 1 : 0);
 
+            var atArrayStart = true;
             while (!valueReader.IsEofDefinition)
             {
-                var defn = valueReader.GetCurrentDefinition();
-
-                if (!firstValue && defn.RepLevel < repetitionLevel)
-                {
-                    break;
-                }
-
-                if (defn.DefLevel == definitionLevel || innerNodeIsOptional)
+                var reachedArrayEnd =
+                    valueReader.ReadValuesAtRepetitionLevel(repetitionLevel, definitionLevel, atArrayStart,
+                        out var valuesSpan);
+                if (reachedArrayEnd && atArrayStart)
                 {
-                    // Note that when the inner node is optional,
-                    // the converter will handle definition levels and create null values.
-                    values.Add(valueReader.ReadValue());
+                    return valuesSpan.ToArray();
                 }
-                else
+                atArrayStart = false;
+                valueChunks.Add(valuesSpan.ToArray());
```

Comment:
Aren't we copying data unnecessarily here? Would it be possible to further improve the performance by using different data type for `valueChunks`, e.g.: `List<(TLogical[] logicalValues, ind index, int size)>`  and avoid creating temporary arrays entirely?

Reply:
No, we need to convert the span to an array before the next call to `ReadValuesAtRepetitionLevel`. The span is pointing into a buffer that will be reused for the next operation, which will overwrite the previously read values.

As long as the array sizes are much smaller than the buffer size (which defaults to 4096 but is configurable), most of the time we should hit the above `reachedArrayEnd && atArrayStart` condition so only do the one copy.